src/stamps.ts

```diff
@@ -1,28 +1,235 @@
+        Number(POSTAGE_DEPTH),
+        POSTAGE_AMOUNT,
+        Number(POSTAGE_USAGE_THRESHOLD || '0.7'),
+        Number(POSTAGE_USAGE_MAX || '0.95'),
+        Number(POSTAGE_TTL_MIN || (refreshPeriod / 1_000) * 5),
+        refreshPeriod,
+        new BeeDebug(BEE_DEBUG_API_URL),
+      )
+    } else if (POSTAGE_DEPTH || POSTAGE_AMOUNT || BEE_DEBUG_API_URL) {
+      throw new Error('Please provide POSTAGE_DEPTH, POSTAGE_AMOUNT and BEE_DEBUG_API_URL environment variable')
+    } else this.enabled = false
+  }
+
+  /**
+   * Get postage stamp that should be replaced in a the proxy request header
+   *
+   * @return Postage stamp that should be used by the proxy
+   *
+   * @throws Error if there is no postage stamp
+   */
+  get postageStamp(): string {
+    if (this.stamp) return this.stamp
+
+    if (this.usableStamps && this.usableStamps[0]) return this.usableStamps[0].batchID
+
+    throw new Error('No postage stamp')
+  }
+
+  /**
+   * Refresh stamps from the bee node and if needed buy new stamp
+   *
+   * @param depth Postage stamps depth
+   * @param amount Postage stamps amount
+   * @param usageThreshold Threshold when new postage stamp should be bought
+   * @param maxUsage Maximal usage of the stamp to be usable by the proxy
+   * @param minTTL Minimal TTL of the stamp to be usable by the proxy
+   * @param beeDebug Connection to debug endpoint for checking/buying stamps
+   */
+  public async refreshStamps(
+    depth: number,
+    amount: string,
+    usageThreshold: number,
+    maxUsage: number,
+    minTTL: number,
+    beeDebug: BeeDebug,
+  ): Promise<void> {
+    try {
+      const stamps = await beeDebug.getAllPostageBatch()
+
+      // Get all usable stamps sorted by usage from most used to least
+      this.usableStamps = filterUsableStamps(stamps, depth, amount, maxUsage, minTTL)
+
+      // Check if the least used stamps is starting to get full and if so purchase new stamp
+      const leastUsed = this.usableStamps[this.usableStamps.length - 1]
+
+      if (!leastUsed || getUsage(leastUsed) > usageThreshold) {
+        buyNewStamp(depth, amount, beeDebug).then(() => {
+          // Once bought, should check if it maybe does not need to be used again
+          this.refreshStamps(depth, amount, usageThreshold, maxUsage, minTTL, beeDebug)
+        })
+      }
+    } catch (e) {
+      logger.error('error: failed to check postage stamp', e)
+    }
+  }
+
+  /**
+   * Start the manager refresh cycle
+   *
+   * @param depth Postage stamps depth
+   * @param amount Postage stamps amount
+   * @param usageThreshold Threshold when new postage stamp should be bought
+   * @param maxUsage Maximal usage of the stamp to be usable by the proxy
+   * @param minTTL Minimal TTL of the stamp to be usable by the proxy
+   * @param refreshPeriod How often should info about stamps be refreshed and potentially new ones be bought
+   * @param beeDebug Connection to debug endpoint for checking/buying stamps
+   */
+  private start(
```

Comment:
Could we make this `async`, `public` and not call this in the constructor but somewhere in the server bootstrap?

What I would like to achieve is that when you start the server the `usableBatches` gets populated before it receives any requests (hence before that the `/rediness` endpoint should be failing), so the `postageStamp` getter should not fail unless something really wrong is happening. There is a possibility that Bee node does not have any stamps (or any meeting the criteria)  so this would then also lead to buying new stamps which is time-consuming process especially with waiting before the stamps are usable.